### PR TITLE
[CUDA graphs] Avoid sync errors when graph capturing cudnn rnn calls that use cudnn dropout

### PR DESCRIPTION
**Pull request by mcarilli**
Cudnn rnn calls that use use cudnn dropout maintain a "state" buffer across calls. [DropoutState](https://github.com/pytorch/pytorch/blob/fe3f6f2da2cb2ddde1a277cd5e99f898933a3c5d/aten/src/ATen/native/cudnn/RNN.cpp#L1388-L1402)'s lock() and unlock() ensure the current call's use of the state buffer syncs with the end of the previous call's use of the state buffer (in case the previous call was on a different stream). 

Telling a capturing stream to wait on an event recorded in a non-capturing stream is an error (1). Telling a non-capturing stream to wait on an event recorded during capture is also an error (2). So DropoutState's flow can error in either of two simple use cases:
```python
rnn = nn.LSTM(512, 512, 2, dropout=0.5).cuda()

out1 = rnn(in1)

# calling cudnn rnn with dropout in capture after calling it uncaptured triggers 1
capture_stream.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(capture_stream):
    graph.capture_begin()
    out2 = rnn(in2)
    graph.capture_end()
torch.cuda.current_stream().wait_stream(capture_stream)

# calling cudnn rnn with dropout uncaptured after calling it in capture triggers 2
out3 = rnn(in3)
```

This PR fixes both cases by telling `DropoutState::lock()`: "if the most recent end-of-usage event was in a different capture state (ie, we crossed a capturing<->noncapturing border) or in a different capture, don't sync on it." While considering the fix I had two assumptions in mind:
- only one capture using the RNN can be underway at a time in this process
- no noncapturing ops in this process are issuing RNN calls while the capture using the RNN is underway.

That second assumption seems brittle if, for example, someone wants to capture an internal region of the forward method of a model wrapped with DataParallel: multiple threads could be issuing RNN calls with some currently capturing and some not. We should talk about whether that use case seems realistic.

(Bigger-picture thoughts: I don't know if forcing calls to serialize on using the shared state buffer is the best design. And if we want to do it that way, we might as well run all cudnn rnns with dropout on a dedicated side stream synced with the surrounding stream (capturing or not), in which case I don't think this PR's event-handling diffs would be needed.)

### CODE REVIEW

**Comment by ezyang** [comment]
Hot take: you should just have your own, separate cache per streaming capture. Then you don't need this PR, and you don't have to worry about the streaming capture stomping all over state that other people have access to.

**Comment by mcarilli** [comment]
> Hot take: you should just have your own, separate cache per streaming capture. Then you don't need this PR, and you don't have to worry about the streaming capture stomping all over state that other people have access to.

That's a good idea to throw into the pot of options, but in that case, to preserve bitwise accuracy of a run mixing uncaptured and capture-replayed rnns with an equivalent fully-uncaptured run, i'd need to copy the eager state to (from) the capture-local state at the beginning (end) of each replay.

**Comment by ezyang** [comment]
You're right, my mistake. (I forgot that this isn't scratch space; it's the RNG state. Blegh, this smells so bad.)

File: aten/src/ATen/native/cudnn/RNN.cpp [hunk 1]
```diff
@@ -1390,13 +1397,38 @@ struct DropoutState {
...
+                                             &status,
+                                             &capture_id_last_unlock));
+      if (status == cudaStreamCaptureStatus::cudaStreamCaptureStatusNone) {
+        capture_id_last_lock = 0;
+      }
+      // Only wait on the last usage if we're not capturing this call in a CUDA graph
+      // or if we're in the same capture as the last usage.
+      if (capture_id_last_lock == capture_id_last_unlock) {
+#endif
+        event->block(cuda::getCurrentCUDAStream());
```
Comment (ezyang) [inline aten/src/ATen/native/cudnn/RNN.cpp:1412]: come on just dupe this code in the else branch lol

**Comment by mcarilli** [comment]
> Blegh, this smells so bad

it's not pretty. At some point i want to have a conversation about what capture usages we intend to allow (eg simultaneous different captures in different threads).

**Comment by ngimel** [comment]
Wait, weren't we going do disallow simultaneous captures already?

**Comment by mcarilli** [comment]
> Wait, weren't we going do disallow simultaneous captures already?

So far i've made a halfassed effort to leave that door open:
The RNG generators currently can't handle it, but maybe they could be tortured into handling it.
The caching allocator diffs, i think, could handle it.

**Comment by ezyang** [comment]
The idea for generators was that if each simultaneous capture had its own, distinct generator, this would probably work out ok (if you hadn't hardcoded a bunch of globals, anyhoo). This is sort of the problem here, which is that you want each capture to have a separate RNG state, if you're actually running this concurrently, but we don't want to pay the cost of this in the public UI.

**Comment by ezyang** [comment]
I guess... I'm OK with it. It probably is well worth putting the commit message in the code, since the correctness conditions are fairly subtle.

**Comment by mcarilli** [comment]
> the commit message

you mean a link to it, or some excerpts?

**Comment by ezyang** [comment]
excerpts; preferably cross referenced as a Note

**Comment by mcarilli** [comment]
The best solution here would be for cudnn to use philox the Right Way like we do (with in-kernel state initialization) so we could lean on Pytorch's existing capturable rng. I can push for that change internally, but I have no idea when or if they'll agree to it.

**Review by ezyang — APPROVED** [review]

**Comment by malfet** [comment]
Reverting the change as it broke OSS CI, see https://app.circleci.com/pipelines/github/pytorch/pytorch/312294/workflows/31e4e524-b6a8-46b3-bd4d-7d883a2d94a4/jobs/12960394

**Pull request closed without merging**